[webkitapipy] Add initial unit tests and hook into test-webkitpy

emw-apple opened the pull request:
#### 49a2255d422e0799beb73477d56db72d74722552
<pre>
[webkitapipy] Add initial unit tests and hook into test-webkitpy
<a href="https://bugs.webkit.org/show_bug.cgi?id=294698">https://bugs.webkit.org/show_bug.cgi?id=294698</a>

Reviewed by Sam Sneddon.

audit-spi gets decent coverage via the iOS build, but as we fix bugs
it&apos;s valuable to be able to write regression tests. Add unit tests to
the library and teach `test-webkitpy` to run them.

To start, contribute tests for the recent windowing logic (#45871) and
ObjC class ingestion (#46874) bug fixes. These are by no means
exhaustive, but seem like a good place to start.

Also add a suite which runs mypy to check the library&apos;s type
annotations. Since changes to type annotations can cause failures at
distant call sites, I believe it&apos;s appropriate to treat them as unit
tests rather than a style check.

* Tools/Scripts/libraries/webkitapipy/mypy.ini: Added. Makes type
  checking succeed in internal builds.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/mypy_unittest.py: Added.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py: Add pragma
  to avoid a typecheck error in open source, when webkitapipy_additions
  is not available.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
(SDKDB.__del__): Support graceful exit when the database has been
  deleted on disk, such as when TestSDKDB deletes the temporary file
  before running `SDKDB.__del__()`.
(SDKDB._add_api_report): Split out from add_partial_sdkdb so that tests
  can call it without invoking the Mach-O reader.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_unittest.py: Added.
* Tools/Scripts/libraries/webkitapipy/webkitapipy_additions.pyi: Added
  type stubs for the internal library.
* Tools/Scripts/webkitpy/test/main.py:
(main): Auto-install mypy and its runtime dependencies.

Canonical link: <a href="https://commits.webkit.org/296534@main">https://commits.webkit.org/296534@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef6f7b45f770e6d6263a87fa04db5461b9f65e40

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108659 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28320 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18744 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113870 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59044 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110622 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29009 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36877 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82543 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111607 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23037 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97879 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62980 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/108091 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22455 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16016 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58577 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92407 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16067 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116991 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35713 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26345 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91563 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36086 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94148 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91367 "Found 2 new API test failures: /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebView/web-process-crashed, /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23300 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36269 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14031 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31546 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35614 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35324 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38670 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37001 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitapipy/mypy.ini:
```diff
@@ -0,0 +1,25 @@
+
+[mypy]
+check_untyped_defs = True
+mypy_path = $MYPY_CONFIG_FILE_DIR/../../../../../Internal/Tools/Scripts/libraries/webkitapipy_additions
```
We presumably want to set `ignore_missing_imports=True` for the `webkitapipy_additions` package? Like:

```ini
[mypy-webkitapipy_additions.*]
ignore_missing_imports = True
```

Or maybe we just provide stubs for it in WebKit, given the public API is inevitably already visible. I think the entire stub we'd want for `webkitapipy_additions.program` is just:

```python
import argparse
from webkitapipy.program import TSVReporter
from webkitapipy.sdkdb import SDKDB as SDKDB

def get_parser() -> argparse.ArgumentParser: ...
def configure_reporter(args: argparse.Namespace, db: SDKDB) -> TSVReporter: ...
```

emw-apple replied:
I like the idea of checking in stubs, and yeah, it's a pretty straightforward interface right now.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/mypy_unittest.py:
```diff
@@ -0,0 +1,36 @@
+
+import mypy.api
+from pathlib import Path
+from unittest import TestCase
```
We mostly follow most Python convention of putting third-party imports in a separate "block" after the stdlib ones, so:

```suggestion
from pathlib import Path
from unittest import TestCase

import mypy.api
```

gsnedders commented on Tools/Scripts/libraries/webkitapipy/mypy.ini:
```diff
@@ -0,0 +1,25 @@
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+[mypy]
```
Can we manage to set `strict = True` here? How far away are we from that being viable?

emw-apple replied:
Still a ways to go (72 errors locally) but none of them look too difficult to address.

gsnedders replied:
We can manage today:

```
strict = True

disallow_any_generics = False
disallow_untyped_calls = False
disallow_untyped_defs = False
disallow_incomplete_defs = False
```

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/mypy_unittest.py:
```diff
@@ -0,0 +1,36 @@
+        module_dir = Path(__file__).parent
+        args = ['--config-file', module_dir.parent / 'mypy.ini',
+                '--pretty', module_dir]
+        stdout, stderr, status = mypy.api.run(list(map(str, args)))
```
`[str(a) for a in args]` might be considered slightly more idiomatic, but this hardly matters.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py:
```diff
@@ -159,7 +159,7 @@ def get_parser() -> argparse.ArgumentParser:
     webkitapipy_additions: Optional[ModuleType]
     try:
-        import webkitapipy_additions
+        import webkitapipy_additions  # type: ignore
```
Does this totally lose us coverage of this, still?

gsnedders commented on Tools/Scripts/webkitpy/test/main.py:
```diff
@@ -86,6 +92,7 @@ def main():
 
     # Tests that are platform specific
     mac_only_tests = (
+        'webkitapipy',
```
FWIW, I have a preference for avoiding adding things here (and actually, this is I think basically unneeded now) — it's much nicer when developing it you can simply run all the tests regardless of host platform!

emw-apple replied:
Yeah, this was a knee-jerk reaction since the tool is only part of Xcode-based ports and analyzes Mach-O binaries, but I think you're right that the tests are (thus far) platform agnostic.

I would like to add some tests in the `macho` module at some point which will regression test parsing of `dyld_info` output, but I can stick to `@unittest.skipIf` decorators for that purpose.

gsnedders replied:
For `dyld_info` output, we'll probably want unit tests we actually control and integration tests we don't — and only the latter are really macOS specific.

gsnedders reviewed (changes requested):
> Since changes to type annotations can cause failures at distant call sites, I believe it's appropriate to treat them as unit tests rather than a style check.

Heh, I remember discussion about this a while ago… maybe it's not the worst option. I feel like we should do better than jamming it in here; would be curious what exact output you get if it fails?

I don't think there's anything major below, but I would be curious as to how far away we are from `strict = True` working; I'm also surprised the missing `webkitapipy_additions` import doesn't cause a failure on EWS.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy_additions.pyi:
```diff
@@ -0,0 +1,31 @@
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+# Annotations for the internal-only webkitapi_additions library.
+# <See https://mypy.readthedocs.io/en/stable/stubs.html>.
```
```suggestion
# See <https://typing.python.org/en/latest/guides/writing_stubs.html>.
```

Maybe a better link than the mypy-specific stuff?

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy_additions.pyi:
```diff
@@ -0,0 +1,31 @@
+from webkitapipy.sdkdb import SDKDB as SDKDB
+
+def get_parser() -> argparse.ArgumentParser: ...
+def configure_reporter(args: argparse.Namespace, db: SDKDB) -> TSVReporter: ...
```
Add a trailing new line to be consistent.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy_additions.pyi:
```diff
@@ -0,0 +1,31 @@
+
+import argparse
+from webkitapipy.program import TSVReporter
+from webkitapipy.sdkdb import SDKDB as SDKDB
```
```suggestion
import argparse

from webkitapipy.program import TSVReporter
from webkitapipy.sdkdb import SDKDB as SDKDB
```

gsnedders commented on Tools/Scripts/libraries/webkitapipy/mypy.ini:
```diff
@@ -0,0 +1,25 @@
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+[mypy]
```
This can actually got in `pyproject.toml` as a `[tool.mypy]` section, and is probably more currently idiomatic.

See e.g. https://mypy.readthedocs.io/en/stable/config_file.html#example-pyproject-toml

emw-apple committed:
[webkitapipy] Add initial unit tests and hook into test-webkitpy
https://bugs.webkit.org/show_bug.cgi?id=294698

Reviewed by Sam Sneddon.

audit-spi gets decent coverage via the iOS build, but as we fix bugs
it's valuable to be able to write regression tests. Add unit tests to
the library and teach `test-webkitpy` to run them.

To start, contribute tests for the recent windowing logic (#45871) and
ObjC class ingestion (#46874) bug fixes. These are by no means
exhaustive, but seem like a good place to start.

Also add a suite which runs mypy to check the library's type
annotations. Since changes to type annotations can cause failures at
distant call sites, I believe it's appropriate to treat them as unit
tests rather than a style check.

* Tools/Scripts/libraries/webkitapipy/mypy.ini: Added. Makes type
  checking succeed in internal builds.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/mypy_unittest.py: Added.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py: Add pragma
  to avoid a typecheck error in open source, when webkitapipy_additions
  is not available.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
(SDKDB.__del__): Support graceful exit when the database has been
  deleted on disk, such as when TestSDKDB deletes the temporary file
  before running `SDKDB.__del__()`.
(SDKDB._add_api_report): Split out from add_partial_sdkdb so that tests
  can call it without invoking the Mach-O reader.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_unittest.py: Added.
* Tools/Scripts/libraries/webkitapipy/webkitapipy_additions.pyi: Added
  type stubs for the internal library.
* Tools/Scripts/webkitpy/test/main.py:
(main): Auto-install mypy and its runtime dependencies.

Canonical link: https://commits.webkit.org/296534@main
